src/nitypes/bintime/_timedelta_array.py

```diff
@@ -105,33 +105,37 @@ def __setitem__(self, index: int | slice, value: TimeDelta | Iterable[TimeDelta]
                 raise invalid_arg_type("value", "TimeDelta", value)
             self._array[index] = value.to_tuple().to_cvi()
         elif isinstance(index, slice):
-            selection = index.indices(len(self))
-            selected_count = len(range(*selection))
-            if selected_count == 0:
-                return
-
             if not isinstance(value, Iterable):
                 raise invalid_arg_type("value", "iterable of TimeDelta", value)
             if not all(isinstance(item, TimeDelta) for item in value):
                 raise invalid_arg_type("value", "iterable of TimeDelta", value)
 
+            start, stop, step = index.indices(len(self))
+            selected_count = len(range(start, stop, step))
             values = list(value)
             new_entry_count = len(values)
-            step_size = selection[-1]
-            if step_size > 1:
-                if new_entry_count <= 1:
-                    raise invalid_arg_value("index", "slice with step size 1", index)
-                elif new_entry_count != selected_count:
-                    raise invalid_arg_value(
-                        "value", "iterable with the same length as the slice", value
-                    )
-
-            if new_entry_count == 0:
-                return self.__delitem__(index)
-            if new_entry_count == 1:
-                self.__delitem__(index)
-                return self.insert(selection[0], values[0])
-            self._array[index] = [item.to_tuple().to_cvi() for item in values]
+            if step > 1 and new_entry_count != selected_count:
+                raise invalid_arg_value(
+                    "value", "iterable with the same length as the slice", value
+                )
+
+            if new_entry_count < selected_count:
+                # Shrink
+                replaced = slice(start, start + new_entry_count)
+                removed = slice(start + new_entry_count, stop)
+                self._array[replaced] = [item.to_tuple().to_cvi() for item in values]
+                del self[removed]
+            elif new_entry_count > selected_count:
+                # Grow
+                replaced = slice(start, stop)
+                self._array[replaced] = [
+                    item.to_tuple().to_cvi() for item in values[:selected_count]
+                ]
+                for offset, entry in enumerate(values[selected_count:]):
+                    self.insert(stop + offset, entry)
```

Comment:
This looks correct, but it shifts the rest of the array rightward N times instead of once.

How about something like this? (untested)
```suggestion
                np.insert(self._array, stop, [item.to_tuple().to_cvi() for item in values[selected_count:])
```